File: Day 1: Print, Variables, and Data Types in Multiple Languages/Day1001.py
# Python Day 1: Print, Variables, and Data Types
# This program demonstrates basic Python concepts including:
# - Print statements
# - Variables
# - All Python data types with their features

# ====================== PRINT STATEMENTS ======================
print("===== PRINT STATEMENTS =====")

# Basic print statement
print("Hello, World!")

# Print multiple items separated by commas (adds spaces automatically)
print("Hello,", "Python", "Learner!")

# Print with custom separator
print("Python", "is", "fun", sep="-")

# Print with custom end character (default is newline)
print("This doesn't end with a newline...", end=" ")
print("See?")

# Formatted string (f-string) - modern way to format output
name = "Python Learner"
print(f"Welcome, {name}!")

# ====================== VARIABLES ======================
print("\n===== VARIABLES =====")

# Variables don't need type declarations
# They are dynamically typed and can change type
x = 10
print(f"x = {x}, type: {type(x)}")

x = "Now I'm a string"
print(f"x = {x}, type: {type(x)}")

# Multiple assignment
a, b, c = 1, 2, 3
print(f"a = {a}, b = {b}, c = {c}")

# Swap variables without a temporary variable
a, b = b, a
print(f"After swap: a = {a}, b = {b}")

# ====================== DATA TYPES ======================
print("\n===== DATA TYPES =====")

# 1. Numeric Types
print("\n-- NUMERIC TYPES --")

# Integer (int) - unlimited precision
integer_number = 42
large_integer = 1234567890123456789012345678901234567890
print(f"Integer: {integer_number}, type: {type(integer_number)}")
print(f"Large integer: {large_integer}")
print(f"Min/Max: Integers in Python 3 have unlimited precision")

# Float - double precision, typically 15-17 significant digits
float_number = 3.14159
scientific_notation = 1.23e-4  # 0.000123
print(f"Float: {float_number}, type: {type(float_number)}")
print(f"Scientific notation: {scientific_notation}")
print(f"Min/Max values: {float('inf')} / {float('-inf')}")
print(f"Special values: NaN (Not a Number): {float('nan')}")

# Complex numbers
complex_number = 3 + 4j
print(f"Complex: {complex_number}, type: {type(complex_number)}")
print(f"Real part: {complex_number.real}, Imaginary part: {complex_number.imag}")
print(f"Conjugate: {complex_number.conjugate()}")

# 2. Boolean Type
print("\n-- BOOLEAN TYPE --")
bool_true = True
bool_false = False
print(f"Boolean values: {bool_true}, {bool_false}, type: {type(bool_true)}")
print(f"Boolean operations: AND: {bool_true and bool_false}, OR: {bool_true or bool_false}, NOT: {not bool_true}")

# 3. Sequence Types
print("\n-- SEQUENCE TYPES --")

# String (str) - immutable sequence of Unicode characters
string_value = "Hello, Python!"
multi_line_string = """This is a
multi-line string"""
print(f"String: {string_value}, type: {type(string_value)}")
print(f"String length: {len(string_value)}")
print(f"String operations: Uppercase: {string_value.upper()}")
print(f"Indexing: {string_value[0]}, {string_value[-1]}")  # First and last character
print(f"Slicing: {string_value[0:5]}")  # First 5 characters
print(f"Multi-line string: {multi_line_string}")

# Raw string (ignores escape characters)
raw_string = r"C:\Users\name\Documents"
print(f"Raw string: {raw_string}")

# List - mutable sequence
my_list = [1, "apple", 3.14, True]
print(f"List: {my_list}, type: {type(my_list)}")
print(f"List length: {len(my_list)}")
print(f"List indexing: {my_list[1]}")
my_list.append("new item")
print(f"After append: {my_list}")
print(f"List comprehension: {[x*2 for x in range(5)]}")

# Tuple - immutable sequence
my_tuple = (1, "apple", 3.14, True)
print(f"Tuple: {my_tuple}, type: {type(my_tuple)}")
print(f"Tuple length: {len(my_tuple)}")
print(f"Tuple indexing: {my_tuple[1]}")
print(f"Single item tuple needs comma: {(1,)}")

# Range - immutable sequence representing a range of numbers
my_range = range(5)  # 0, 1, 2, 3, 4
print(f"Range: {my_range}, type: {type(my_range)}")
print(f"Range as list: {list(my_range)}")
print(f"Range with start, stop, step: {list(range(2, 10, 2))}")

# 4. Mapping Type
print("\n-- MAPPING TYPE --")

# Dictionary (dict) - mutable mapping of keys to values
my_dict = {"name": "Python", "year": 1991, "is_awesome": True}
print(f"Dictionary: {my_dict}, type: {type(my_dict)}")
print(f"Dictionary access: {my_dict['name']}")
print(f"Dictionary get (safe): {my_dict.get('version', 'Not found')}")
my_dict["version"] = 3.10
print(f"After adding a key: {my_dict}")
print(f"Dictionary keys: {list(my_dict.keys())}")
print(f"Dictionary values: {list(my_dict.values())}")
print(f"Dictionary items: {list(my_dict.items())}")

# 5. Set Types
print("\n-- SET TYPES --")

# Set - mutable unordered collection of unique elements
my_set = {1, 2, 3, 2, 1}  # Duplicates are automatically removed
print(f"Set: {my_set}, type: {type(my_set)}")
print(f"Set length: {len(my_set)}")
my_set.add(4)
print(f"After adding element: {my_set}")
print(f"Set operations: Union: {my_set | {5, 6}}, Intersection: {my_set & {3, 4, 5}}")

# Frozen set - immutable version of set
my_frozen_set = frozenset([1, 2, 3, 2, 1])
print(f"Frozen set: {my_frozen_set}, type: {type(my_frozen_set)}")
print(f"Frozen set is immutable and hashable")

# 6. Binary Types
print("\n-- BINARY TYPES --")

# Bytes - immutable sequence of bytes
my_bytes = b"hello"
print(f"Bytes: {my_bytes}, type: {type(my_bytes)}")
print(f"Bytes indexing gives integers: {my_bytes[0]}")

# Bytearray - mutable sequence of bytes
my_bytearray = bytearray(b"hello")
print(f"Bytearray: {my_bytearray}, type: {type(my_bytearray)}")
my_bytearray[0] = 72  # ASCII for 'H'
print(f"After modification: {my_bytearray}")

# Memoryview - memory view of a bytes-like object
my_memoryview = memoryview(b"hello")
print(f"Memoryview: {my_memoryview}, type: {type(my_memoryview)}")
print(f"Memoryview access: {my_memoryview[0]}")

# 7. None Type
print("\n-- NONE TYPE --")
my_none = None
print(f"None value: {my_none}, type: {type(my_none)}")
print(f"None is used to represent the absence of a value")

# 8. Special Collections (from collections module)
print("\n-- SPECIAL COLLECTIONS --")

# Named Tuple
from collections import namedtuple
Point = namedtuple('Point', ['x', 'y'])
p = Point(11, y=22)
print(f"Named tuple: {p}, type: {type(p)}")
print(f"Named tuple access: {p.x}, {p[1]}")

# OrderedDict (note: regular dicts are now ordered by default in Python 3.7+)
from collections import OrderedDict
ordered_dict = OrderedDict([('a', 1), ('b', 2)])
print(f"OrderedDict: {ordered_dict}, type: {type(ordered_dict)}")

# Counter
from collections import Counter
my_counter = Counter("mississippi")
print(f"Counter: {my_counter}, type: {type(my_counter)}")
print(f"Most common: {my_counter.most_common(2)}")

# Deque (double-ended queue)
from collections import deque
my_deque = deque([1, 2, 3])
print(f"Deque: {my_deque}, type: {type(my_deque)}")
my_deque.appendleft(0)
my_deque.append(4)
print(f"After modification: {my_deque}")

# DefaultDict
from collections import defaultdict
my_defaultdict = defaultdict(int)  # Default factory returns 0 for missing keys
my_defaultdict['a'] += 1
print(f"DefaultDict: {dict(my_defaultdict)}, type: {type(my_defaultdict)}")
print(f"Missing key returns: {my_defaultdict['b']}")

# 9. Decimal and Fraction (for precise arithmetic)
print("\n-- DECIMAL AND FRACTION --")

# Decimal - precise decimal arithmetic
from decimal import Decimal
my_decimal = Decimal('0.1')
print(f"Decimal: {my_decimal}, type: {type(my_decimal)}")
print(f"Regular float 0.1 + 0.2: {0.1 + 0.2}")  # Shows floating point precision issues
print(f"Decimal 0.1 + 0.2: {Decimal('0.1') + Decimal('0.2')}")  # Precise

# Fraction - rational numbers
from fractions import Fraction
my_fraction = Fraction(1, 3)
print(f"Fraction: {my_fraction}, type: {type(my_fraction)}")
print(f"Fraction from float: {Fraction(1.5)}")
print(f"Fraction from string: {Fraction('1.5')}")

# 10. Datetime Types
print("\n-- DATETIME TYPES --")

# DateTime, Date, Time
from datetime import datetime, date, time, timedelta
current_datetime = datetime.now()
my_date = date(2025, 2, 24)
my_time = time(13, 30, 0)
print(f"Datetime: {current_datetime}, type: {type(current_datetime)}")
print(f"Date: {my_date}, type: {type(my_date)}")
print(f"Time: {my_time}, type: {type(my_time)}")

# TimeDelta - represents a duration
one_day = timedelta(days=1)
tomorrow = my_date + one_day
print(f"TimeDelta: {one_day}, type: {type(one_day)}")
print(f"Tomorrow: {tomorrow}")

# 11. Enum Type (from enum module)
print("\n-- ENUM TYPE --")

from enum import Enum, auto
class Color(Enum):
    RED = 1
    GREEN = 2
    BLUE = auto()  # auto-assigns the next value

print(f"Enum member: {Color.RED}, type: {type(Color.RED)}")
print(f"Enum value: {Color.RED.value}, name: {Color.RED.name}")
print(f"Auto value: {Color.BLUE.value}")

# Type Conversion Examples
print("\n===== TYPE CONVERSION =====")
num_int = 123
num_float = float(num_int)
num_str = str(num_int)
num_bool = bool(num_int)  # Any non-zero number is True
str_list = list("hello")  # Convert string to list of characters

print(f"Original int: {num_int}")
print(f"Converted to float: {num_float}")
print(f"Converted to string: {num_str}")
print(f"Converted to bool: {num_bool}")
print(f"String to list: {str_list}")

# Finding the type and checking if an object is of a certain type
print("\n===== TYPE CHECKING =====")
value = 42
print(f"Type of value: {type(value)}")
print(f"Is integer? {isinstance(value, int)}")
print(f"Is numeric? {isinstance(value, (int, float, complex))}")

# Memory size of objects
print("\n===== OBJECT SIZE =====")
import sys
print(f"Size of integer: {sys.getsizeof(0)} bytes")
print(f"Size of empty string: {sys.getsizeof('')} bytes")
print(f"Size of empty list: {sys.getsizeof([])} bytes")
print(f"Size of empty dict: {sys.getsizeof({})} bytes")

File: Day 1: Print, Variables, and Data Types in Multiple Languages/Day1001.cpp
// C++ Day 1: Print, Variables, and Data Types

#include <iostream>     // Required for input/output operations
#include <string>       // Required for string operations
#include <limits>       // Required for checking min/max values of data types
#include <climits>      // Required for integer limits
#include <cfloat>       // Required for floating-point limits
#include <complex>      // Required for complex numbers
#include <vector>       // Required for vectors
#include <array>        // Required for arrays
#include <list>         // Required for linked lists
#include <map>          // Required for maps
#include <set>          // Required for sets
#include <queue>        // Required for queues
#include <stack>        // Required for stacks
#include <bitset>       // Required for bitsets
#include <memory>       // Required for smart pointers

int main() {
    // Print statements - displaying output to the console
    std::cout << "Hello, World!" << std::endl;  // Simple string output
    std::cout << "Welcome to C++ Programming!" << std::endl;
    
    // Variables - containers for storing data values
    std::string name = "Alice";  // String variable
    int age = 25;                // Integer variable
    double height = 5.7;         // Double variable
    bool isStudent = true;       // Boolean variable
    
    std::cout << "\nVariable demonstration:" << std::endl;
    std::cout << "Name: " << name << std::endl;
    std::cout << "Age: " << age << std::endl;
    std::cout << "Height: " << height << " feet" << std::endl;
    std::cout << "Is student? " << std::boolalpha << isStudent << std::endl;  // std::boolalpha shows true/false instead of 1/0
    
    // Multiple variable assignment
    int x = 10, y = 20, z = 30;
    std::cout << "\nMultiple variables: " << x << " " << y << " " << z << std::endl;
    
    // --------------------------------------------------------
    // C++ DATA TYPES - COMPREHENSIVE OVERVIEW
    // --------------------------------------------------------
    
    std::cout << "\n=== C++ DATA TYPES ===" << std::endl;
    
    // 1. INTEGER TYPES
    std::cout << "\n--- INTEGER TYPES ---" << std::endl;
    
    // Character (char) - typically 1 byte
    char character = 'A';
    std::cout << "\nChar: " << character << std::endl;
    std::cout << "Size of char: " << sizeof(char) << " bytes" << std::endl;
    std::cout << "Char as integer: " << static_cast<int>(character) << std::endl;
    std::cout << "Min char value: " << static_cast<int>(CHAR_MIN) << std::endl;
    std::cout << "Max char value: " << static_cast<int>(CHAR_MAX) << std::endl;
    
    // Signed char - guaranteed to be 1 byte
    signed char signedChar = -128;
    std::cout << "\nSigned char: " << static_cast<int>(signedChar) << std::endl;
    std::cout << "Size of signed char: " << sizeof(signed char) << " bytes" << std::endl;
    std::cout << "Min signed char value: " << static_cast<int>(SCHAR_MIN) << std::endl;
    std::cout << "Max signed char value: " << static_cast<int>(SCHAR_MAX) << std::endl;
    
    // Unsigned char - guaranteed to be 1 byte
    unsigned char unsignedChar = 255;
    std::cout << "\nUnsigned char: " << static_cast<int>(unsignedChar) << std::endl;
    std::cout << "Size of unsigned char: " << sizeof(unsigned char) << " bytes" << std::endl;
    std::cout << "Min unsigned char value: 0" << std::endl;
    std::cout << "Max unsigned char value: " << static_cast<int>(UCHAR_MAX) << std::endl;
    
    // Wide character (wchar_t)
    wchar_t wideChar = L'Ω';  // Unicode character
    std::cout << "\nWide character (as number): " << static_cast<int>(wideChar) << std::endl;
    std::cout << "Size of wchar_t: " << sizeof(wchar_t) << " bytes" << std::endl;
    std::wcout << L"Wide character: " << wideChar << std::endl;
    
    // Char16_t (C++11) - for UTF-16 characters
    char16_t char16 = u'Ω';
    std::cout << "\nChar16_t (as number): " << static_cast<int>(char16) << std::endl;
    std::cout << "Size of char16_t: " << sizeof(char16_t) << " bytes" << std::endl;
    
    // Char32_t (C++11) - for UTF-32 characters
    char32_t char32 = U'Ω';
    std::cout << "\nChar32_t (as number): " << static_cast<int>(char32) << std::endl;
    std::cout << "Size of char32_t: " << sizeof(char32_t) << " bytes" << std::endl;
    
    // Short int - typically 2 bytes
    short shortNum = 32767;
    std::cout << "\nShort int: " << shortNum << std::endl;
    std::cout << "Size of short: " << sizeof(short) << " bytes" << std::endl;
    std::cout << "Min short value: " << SHRT_MIN << std::endl;
    std::cout << "Max short value: " << SHRT_MAX << std::endl;
    
    // Unsigned short int
    unsigned short unsignedShort = 65535;
    std::cout << "\nUnsigned short: " << unsignedShort << std::endl;
    std::cout << "Size of unsigned short: " << sizeof(unsigned short) << " bytes" << std::endl;
    std::cout << "Min unsigned short value: 0" << std::endl;
    std::cout << "Max unsigned short value: " << USHRT_MAX << std::endl;
    
    // Integer (int) - typically 4 bytes
    int integerNum = 2147483647;
    std::cout << "\nInt: " << integerNum << std::endl;
    std::cout << "Size of int: " << sizeof(int) << " bytes" << std::endl;
    std::cout << "Min int value: " << INT_MIN << std::endl;
    std::cout << "Max int value: " << INT_MAX << std::endl;
    
    // Unsigned int
    unsigned int unsignedInt = 4294967295;
    std::cout << "\nUnsigned int: " << unsignedInt << std::endl;
    std::cout << "Size of unsigned int: " << sizeof(unsigned int) << " bytes" << std::endl;
    std::cout << "Min unsigned int value: 0" << std::endl;
    std::cout << "Max unsigned int value: " << UINT_MAX << std::endl;
    
    // Long int - at least 4 bytes
    long longNum = 2147483647L;
    std::cout << "\nLong int: " << longNum << std::endl;
    std::cout << "Size of long: " << sizeof(long) << " bytes" << std::endl;
    std::cout << "Min long value: " << LONG_MIN << std::endl;
    std::cout << "Max long value: " << LONG_MAX << std::endl;
    
    // Unsigned long int
    unsigned long unsignedLong = 4294967295UL;
    std::cout << "\nUnsigned long: " << unsignedLong << std::endl;
    std::cout << "Size of unsigned long: " << sizeof(unsigned long) << " bytes" << std::endl;
    std::cout << "Min unsigned long value: 0" << std::endl;
    std::cout << "Max unsigned long value: " << ULONG_MAX << std::endl;
    
    // Long long int (C++11) - at least 8 bytes
    long long longLongNum = 9223372036854775807LL;
    std::cout << "\nLong long int: " << longLongNum << std::endl;
    std::cout << "Size of long long: " << sizeof(long long) << " bytes" << std::endl;
    std::cout << "Min long long value: " << LLONG_MIN << std::endl;
    std::cout << "Max long long value: " << LLONG_MAX << std::endl;
    
    // Unsigned long long int (C++11)
    unsigned long long unsignedLongLong = 18446744073709551615ULL;
    std::cout << "\nUnsigned long long: " << unsignedLongLong << std::endl;
    std::cout << "Size of unsigned long long: " << sizeof(unsigned long long) << " bytes" << std::endl;
    std::cout << "Min unsigned long long value: 0" << std::endl;
    std::cout << "Max unsigned long long value: " << ULLONG_MAX << std::endl;
    
    // 2. FLOATING-POINT TYPES
    std::cout << "\n--- FLOATING-POINT TYPES ---" << std::endl;
    
    // Float - typically 4 bytes, 7 digits precision
    float floatNum = 3.14159f;
    std::cout << "\nFloat: " << floatNum << std::endl;
    std::cout << "Size of float: " << sizeof(float) << " bytes" << std::endl;
    std::cout << "Float digits of precision: " << FLT_DIG << std::endl;
    std::cout << "Min positive float value: " << FLT_MIN << std::endl;
    std::cout << "Max float value: " << FLT_MAX << std::endl;
    
    // Double - typically 8 bytes, 15 digits precision
    double doubleNum = 3.1415926535897932;
    std::cout << "\nDouble: " << doubleNum << std::endl;
    std::cout << "Size of double: " << sizeof(double) << " bytes" << std::endl;
    std::cout << "Double digits of precision: " << DBL_DIG << std::endl;
    std::cout << "Min positive double value: " << DBL_MIN << std::endl;
    std::cout << "Max double value: " << DBL_MAX << std::endl;
    
    // Long double - typically 12 or 16 bytes
    long double longDoubleNum = 3.1415926535897932384626433L;
    std::cout << "\nLong double: " << longDoubleNum << std::endl;
    std::cout << "Size of long double: " << sizeof(long double) << " bytes" << std::endl;
    std::cout << "Long double digits of precision: " << LDBL_DIG << std::endl;
    std::cout << "Min positive long double value: " << LDBL_MIN << std::endl;
    std::cout << "Max long double value: " << LDBL_MAX << std::endl;
    
    // 3. COMPLEX NUMBER TYPES
    std::cout << "\n--- COMPLEX NUMBER TYPES ---" << std::endl;
    
    // Complex float
    std::complex<float> complexFloat(3.0f, 4.0f);
    std::cout << "\nComplex float: " << complexFloat << std::endl;
    std::cout << "Size of complex<float>: " << sizeof(std::complex<float>) << " bytes" << std::endl;
    std::cout << "Real part: " << complexFloat.real() << std::endl;
    std::cout << "Imaginary part: " << complexFloat.imag() << std::endl;
    std::cout << "Absolute value (magnitude): " << std::abs(complexFloat) << std::endl;
    
    // Complex double
    std::complex<double> complexDouble(3.0, 4.0);
    std::cout << "\nComplex double: " << complexDouble << std::endl;
    std::cout << "Size of complex<double>: " << sizeof(std::complex<double>) << " bytes" << std::endl;
    
    // 4. BOOLEAN TYPE
    std::cout << "\n--- BOOLEAN TYPE ---" << std::endl;
    
    // Boolean (bool)
    bool boolTrue = true;
    bool boolFalse = false;
    std::cout << "\nBool (true): " << std::boolalpha << boolTrue << std::endl;
    std::cout << "Bool (false): " << boolFalse << std::endl;
    std::cout << "Size of bool: " << sizeof(bool) << " bytes" << std::endl;
    
    // 5. VOID TYPE
    std::cout << "\n--- VOID TYPE ---" << std::endl;
    std::cout << "Void is a special type that indicates 'no value'" << std::endl;
    std::cout << "It's commonly used for functions that don't return a value" << std::endl;
    
    // Void pointer - can point to any data type
    void* voidPtr = &age;
    std::cout << "\nVoid pointer address: " << voidPtr << std::endl;
    
    // 6. NULL POINTER
    std::cout << "\n--- NULL POINTER ---" << std::endl;
    
    // Null pointer (C-style)
    int* nullPtr = NULL;
    std::cout << "Null pointer (NULL): " << nullPtr << std::endl;
    
    // Null pointer (C++11)
    int* nullPtr2 = nullptr;
    std::cout << "Null pointer (nullptr): " << nullPtr2 << std::endl;
    
    // 7. STRING TYPES
    std::cout << "\n--- STRING TYPES ---" << std::endl;
    
    // C-style string (character array)
    char cString[] = "Hello C++";
    std::cout << "\nC-style string: " << cString << std::endl;
    std::cout << "Size: " << sizeof(cString) << " bytes (including null terminator)" << std::endl;
    
    // std::string - modern C++ string
    std::string stdString = "Hello modern C++";
    std::cout << "\nstd::string: " << stdString << std::endl;
    std::cout << "Length: " << stdString.length() << " characters" << std::endl;
    std::cout << "Capacity: " << stdString.capacity() << " characters" << std::endl;
    std::cout << "First character: " << stdString[0] << std::endl;
    std::cout << "Substring: " << stdString.substr(6, 6) << std::endl;
    
    // String concatenation
    std::string firstName = "John";
    std::string lastName = "Doe";
    std::string fullName = firstName + " " + lastName;
    std::cout << "\nConcatenated string: " << fullName << std::endl;
    
    // Wide string (wstring)
    std::wstring wideString = L"Wide string with Unicode: Ω";
    std::wcout << L"\nWide string (wstring): " << wideString << std::endl;
    
    // 8. CONTAINER TYPES FROM STL
    std::cout << "\n--- CONTAINER TYPES (STL) ---" << std::endl;
    
    // Array (fixed-size array)
    std::array<int, 5> stdArray = {1, 2, 3, 4, 5};
    std::cout << "\nstd::array: ";
    for (const auto& element : stdArray) {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    std::cout << "Size: " << stdArray.size() << std::endl;
    std::cout << "First element: " << stdArray.front() << std::endl;
    std::cout << "Last element: " << stdArray.back() << std::endl;
    
    // Vector (dynamic array)
    std::vector<int> stdVector = {10, 20, 30, 40, 50};
    std::cout << "\nstd::vector: ";
    for (const auto& element : stdVector) {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    std::cout << "Size: " << stdVector.size() << std::endl;
    std::cout << "Capacity: " << stdVector.capacity() << std::endl;
    stdVector.push_back(60);
    std::cout << "After push_back: ";
    for (const auto& element : stdVector) {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    
    // List (doubly-linked list)
    std::list<int> stdList = {100, 200, 300};
    std::cout << "\nstd::list: ";
    for (const auto& element : stdList) {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    stdList.push_front(50);
    stdList.push_back(400);
    std::cout << "After push operations: ";
    for (const auto& element : stdList) {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    
    // Map (associative array)
    std::map<std::string, int> stdMap;
    stdMap["one"] = 1;
    stdMap["two"] = 2;
    stdMap["three"] = 3;
    std::cout << "\nstd::map: " << std::endl;
    for (const auto& pair : stdMap) {
        std::cout << pair.first << ": " << pair.second << std::endl;
    }
    std::cout << "Access by key (\"two\"): " << stdMap["two"] << std::endl;
    
    // Set (collection of unique values)
    std::set<int> stdSet = {10, 20, 30, 10, 20};  // Duplicates are ignored
    std::cout << "\nstd::set: ";
    for (const auto& element : stdSet) {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    stdSet.insert(40);
    std::cout << "After insert: ";
    for (const auto& element : stdSet) {
        std::cout << element << " ";
    }
    std::cout << std::endl;

    // Queue (FIFO data structure)
    std::queue<int> stdQueue;
    stdQueue.push(100);
    stdQueue.push(200);
    stdQueue.push(300);
    std::cout << "\nstd::queue (front): " << stdQueue.front() << std::endl;
    std::cout << "std::queue (back): " << stdQueue.back() << std::endl;
    std::cout << "Size: " << stdQueue.size() << std::endl;
    stdQueue.pop();  // Removes the front element
    std::cout << "After pop, new front: " << stdQueue.front() << std::endl;
    
    // Stack (LIFO data structure)
    std::stack<int> stdStack;
    stdStack.push(100);
    stdStack.push(200);
    stdStack.push(300);
    std::cout << "\nstd::stack (top): " << stdStack.top() << std::endl;
    std::cout << "Size: " << stdStack.size() << std::endl;
    stdStack.pop();  // Removes the top element
    std::cout << "After pop, new top: " << stdStack.top() << std::endl;
    
    // 9. SPECIAL TYPES
    std::cout << "\n--- SPECIAL TYPES ---" << std::endl;
    
    // Enum (enumeration)
    enum Color { RED, GREEN, BLUE };  // By default, RED=0, GREEN=1, BLUE=2
    Color myColor = GREEN;
    std::cout << "\nEnum value: " << myColor << std::endl;
    
    // Enum class (C++11, strongly typed enumeration)
    enum class Direction { NORTH, EAST, SOUTH, WEST };
    Direction myDirection = Direction::EAST;
    std::cout << "Enum class used (cannot display raw value directly)" << std::endl;
    
    // Bitset
    std::bitset<8> bits(42);  // 00101010 in binary
    std::cout << "\nBitset (decimal 42): " << bits << std::endl;
    std::cout << "Bit at position 1: " << bits[1] << std::endl;
    std::cout << "Number of set bits: " << bits.count() << std::endl;
    bits.flip();  // Flip all bits
    std::cout << "After flip: " << bits << std::endl;
    
    // 10. SMART POINTERS (C++11)
    std::cout << "\n--- SMART POINTERS ---" << std::endl;
    
    // Unique pointer - exclusive ownership
    std::unique_ptr<int> uniquePtr = std::make_unique<int>(42);  // C++14
    std::cout << "\nUnique pointer value: " << *uniquePtr << std::endl;
    
    // Shared pointer - shared ownership
    std::shared_ptr<int> sharedPtr1 = std::make_shared<int>(100);
    std::shared_ptr<int> sharedPtr2 = sharedPtr1;  // Both point to the same memory
    std::cout << "\nShared pointer value: " << *sharedPtr1 << std::endl;
    std::cout << "Reference count: " << sharedPtr1.use_count() << std::endl;
    
    // Weak pointer - non-owning reference to a shared_ptr
    std::weak_ptr<int> weakPtr = sharedPtr1;
    std::cout << "\nWeak pointer created (doesn't affect reference count)" << std::endl;
    std::cout << "Reference count: " << sharedPtr1.use_count() << std::endl;
    
    // 11. TYPE INFERENCE (C++11)
    std::cout << "\n--- TYPE INFERENCE ---" << std::endl;
    
    // auto - automatically deduce variable type
    auto autoInt = 42;
    auto autoDouble = 3.14159;
    auto autoString = "auto string";
    
    std::cout << "\nAuto int: " << autoInt << std::endl;
    std::cout << "Auto double: " << autoDouble << std::endl;
    std::cout << "Auto string: " << autoString << std::endl;
    
    // decltype - get the type of an expression
    int num = 42;
    decltype(num) declInt = 100;  // Same type as num (int)
    decltype(3.14159) declDouble = 2.71828;  // double
    
    std::cout << "\nDecltype int: " << declInt << std::endl;
    std::cout << "Decltype double: " << declDouble << std::endl;
    
    // 12. TYPE CONVERSIONS
    std::cout << "\n--- TYPE CONVERSIONS ---" << std::endl;
    
    // Implicit conversion
    int intVal = 42;
    double doubleVal = intVal;  // int to double (no data loss)
    std::cout << "\nImplicit int to double: " << doubleVal << std::endl;
    
    // Potential data loss
    double piVal = 3.14159;
    int piInt = piVal;  // double to int (loss of fractional part)
    std::cout << "Double to int (data loss): " << piInt << std::endl;
    
    // C-style casting
    float floatVal = 2.71828f;
    int castInt = (int)floatVal;
    std::cout << "\nC-style cast (float to int): " << castInt << std::endl;
    
    // Modern C++ casting
    
    // static_cast - compile-time type checking
    double doubleVal2 = 3.14159;
    int staticCastInt = static_cast<int>(doubleVal2);
    std::cout << "\nstatic_cast (double to int): " << staticCastInt << std::endl;
    
    // reinterpret_cast - low-level reinterpretation of bit patterns
    int* intPtr = new int(42);
    long longFromPtr = reinterpret_cast<long>(intPtr);
    std::cout << "reinterpret_cast (pointer to long): " << longFromPtr << std::endl;
    int* ptrBack = reinterpret_cast<int*>(longFromPtr);
    std::cout << "reinterpret_cast (long back to pointer), value: " << *ptrBack << std::endl;
    
    // const_cast - add or remove const qualifier
    const int constVal = 100;
    int& nonConstRef = const_cast<int&>(constVal);  // Removing const (potentially dangerous)
    std::cout << "\nconst_cast (removing const): " << nonConstRef << std::endl;
    
    // dynamic_cast - safe downcast for polymorphic types
    // Used with inheritance, requires runtime type information
    
    // Memory cleanup
    delete intPtr;
    
    // 13. USER-DEFINED TYPES
    std::cout << "\n--- USER-DEFINED TYPES ---" << std::endl;
    
    // Struct - grouping related data
    struct Person {
        std::string name;
        int age;
        double height;
    };
    
    Person person1 = {"John", 30, 5.9};
    std::cout << "\nStruct Person: " << person1.name << ", " << person1.age << " years, " 
              << person1.height << " feet" << std::endl;
    
    // Union - memory-saving way to store different data types (one at a time)
    union Value {
        int intValue;
        double doubleValue;
        char charValue;
    };
    
    Value value;
    value.intValue = 42;
    std::cout << "\nUnion with int: " << value.intValue << std::endl;
    value.doubleValue = 3.14159;  // Overwrites the int value
    std::cout << "Union with double (now): " << value.doubleValue << std::endl;
    std::cout << "Union int value (corrupted): " << value.intValue << std::endl;
    
    // Class - the basis for OOP in C++
    std::cout << "\nClass (object-oriented programming):" << std::endl;
    std::cout << "A complex data type that encapsulates data and methods" << std::endl;
    
    return 0;
}